### ISSUE TITLE
Set up Daily Manifests for Post Ingest, Continuous, Deletes of MAXAR MCP Delivery Bucket Files

### ISSUE THREAD

**Issue by krisstanton** [NASA-IMPACT/csdap-cumulus#328]
Update: The scope of this ticket has been changed to provide manifest files from NGAP to MCP for the purpose of setting up continuous deletion.

General Background:
After Cumulus runs, we have the following.
- files in the CBA PROD bucket(s) (Protected and Public)
- files in the CBA PROD DR bucket (Archive) (Note, these files are sourced from BOTH Protected and Public CBA Prod buckets and kind of smashed together)
- CMR Data published on Earthdata
- External Metrics confirmations
S3 Bucket directory and file listing can be acquired by manifest lists generated by S3.

This ticket deals with getting those manifest files from NGAP over to MCP.  All processing of this info will be done on MCP by an Archive DAG

Tasks:
- [x] Setup MCP to be able to receive inventories
  - [x] Create destination directories in bucket `ss-s3-inventories-uswest2`
    - [x] CBA PROD Protected: `ss-s3-inventories-uswest2/ngap-buckets/cba-prod-protected/`
    - [x] CBA PROD Public: `ss-s3-inventories-uswest2/ngap-buckets/cba-prod-public/`
    - [x] CBA PROD ORCA: `ss-s3-inventories-uswest2/ngap-buckets/cba-prod-orca/`
  - [x] Create Role on MCP `csda-ngap-inventory-to-s3-writer-cba-prod`
    - [x] Attach Policy for writing to the specific bucket and subdirectory `ss-s3-inventories-uswest2/ngap-buckets/*`
  - [x] Create Role on MCP `csda-ngap-inventory-to-s3-writer-cba-prod-orca`
    - [x] Attach Policy for writing to the specific bucket and subdirectory `ss-s3-inventories-uswest2/ngap-buckets/*`
  - [x] Attach Policy to the MCP bucket that allows writing from both CBA PROD AWS accounts
- [x] Send the Manifest files from NGAP (CBA PROD) to the MCP Bucket, `ss-s3-inventories-uswest2` in the correct subdirectory
- [x] Send the Manifest files from ORCA (CBA PROD DR) to the MCP Bucket, `ss-s3-inventories-uswest2` in the correct subdirectory
- [x] Verify the inventory files are in their expected locations on MCP
- Make sure we have a way to auto remove these files in 3 months (lifecycle rule? TTL? etc) for all 3 of these inventories  UPDATE: For now, this point on 3 month lifecycle is part of a greater discussion on how our DMS system will handle any and all lifecycle rules of the incoming inventories.  So this part of the task is shelved until that larger discussion is resolved.

**Everything below is old / archived / draft while coming up with an approach to solving this item **

***
Preserving the older description again (Immediately Below) for this ticket so that we ensure nothing is missed
***

This ticket is being repurposed to be in line with using our Data Management System (DMS) in order to automate the process of doing continuous deletions.

The scope of this ticket will be limited to Prototyping work immediately following an agreement on how we are approaching this problem.
Reference to Scoping Ticket: https://github.com/NASA-IMPACT/csdap-cumulus/issues/367
Reference to Milestone: https://github.com/NASA-IMPACT/csda-project/issues/651

Please Edit the below steps to include new Items discussed in the approach.
Also remove or modify below steps to reflect any changes that come from those discussions.

Steps (DRAFT)
- [ ] Create a python process which can be easily converted or adapted to be a DAG which does the following
  - [ ] Examine the data from the perspective of specific granules
    - [ ] Makes CMR Request for a specific Granule
      - [ ] Validates the source of S3 paths to actually be pointing to CBA PROD (Remember, some of the published data still points to OLD NGAP)
    - [ ] Makes request to read data in CBA PROD Buckets and CBA DR PROD (ORCA) Buckets
    - [ ] Uses the data returned to determine if an individual granule (or batch of granules) exists in all 3 places
  - [ ] Process Manifests (Possible Scaling Problem here??)
    - [ ] Read a batch of files from CBA PROD's Manifest
    - [ ] Verify those granules (collection of files) exist in ORCA (CBA PROD DR account)
    - [ ] Verify those granules are published in the CMR Record with correct paths. 
Note: Some of this code already exists, but it may need to be set up to process in batches (so new controlling structure in the code may be needed).

***
Preserving the older description (Immediately Below) for this ticket so that readily developed tools for this task can be more easily found and referenced
***

OLD Name: MCP Manifest Analysis: Utility to generate Lists: SAFE_TO_DELETE', and 'STILL_NEED_TO_INGEST'
OLD DESCRIPTION BELOW

MCP Manifest Lists Generation

Update: Important Note: We will be deleting triple verified (Earthdata, CBA and CBA ORCA) files from NGAP FIRST, before we delete anything from MCP.

This is a utility that runs AFTER the ORCA Validation Utility AND the NGAP List Generator (this is part 3 of 3 of the Cumulus Pipeline Manifest utilities).

The functionality here is very similar to that found in https://github.com/NASA-IMPACT/csdap-cumulus/issues/319

The purpose of this is to generate two lists: SAFE_TO_DELETE, STILL_NEED_TO_INGEST for both MCP.
This is done by examining the MCP Maxar Delivery Bucket Manifest and comparing it with the outputs from the ORCA Validation Utility (https://github.com/NASA-IMPACT/csdap-cumulus/issues/270).

Note: The file names should be the same but the paths to them might be different (I believe the discrepancy is between MCP Maxar Delivery Bucket Paths and the NGAP/CBA Bucket Paths).  We may need some additional logic in the utility that compares only parts of the file path (instead of the entire file path) from different manifests to generate these lists.

Files in the MCP: STILL_NEED_TO_INGEST list here will require a run through Airflow (restore and convert xml to cmr) and then Cumulus
Files in the NGAP: STILL_NEED_TO_INGEST list here would only require a run through Airflow from the MCP bucket IF the CMR record is missing (there were a small number of these cases).  The default case for these files would be to just run the Cumulus ingest rule which covers these items.

**Comment by krisstanton** [comment]
Update: Important Note (adding this to description as well) - We will be deleting files from NGAP FIRST, BEFORE deleting anything from MCP.

**Comment by krisstanton** [comment]
WIP UPDATE:
After some conversations and research.  Here are the initial steps:

The destination bucket for the inventory files will be MCP Account:  `ss-s3-inventories-uswest2`
With Subdirectory: `ngap-buckets`
There are 3 sets of inventories that need to be sent from 2 different AWS accounts to MCP, here are the full paths and descriptions.
```
// This bucket contains granule content files that are secured behind the cognito login
// // This bucket contains all datasets that Cumulus Publishes (MAXAR and PlanetScope Data)
From CBA PROD (protected-bucket) 	---> MCP_Account ss-s3-inventories-uswest2/ngap-buckets/cba-prod-protected/

// This bucket contains only Thumbnails which are publicly available (so earth data can directly reference them per granule)
// // This bucket contains all datasets that Cumulus Publishes (MAXAR and PlanetScope Data)
From CBA PROD (public-bucket) 		---> MCP_Account ss-s3-inventories-uswest2/ngap-buckets/cba-prod-public/

// This bucket contains both Thumbnails and Granules files that are backed up.
// // NOTE: Orca Archive Bucket smashes the protected and public buckets together, but really it's the key path that matters 
// // // (On a side tangent Note: ORCA has it's own way to restore back to original paths - so this mixing of buckets is really ok)
From CBA PROD ORCA (archive-bucket) ---> MCP_Account ss-s3-inventories-uswest2/ngap-buckets/cba-prod-orca/
```

Need to make new roles on MCP which can be used by NGAP to write to these buckets.
```
new role:  csda-ngap-inventory-to-s3-writer-cba-prod 				// csda-cumulus-prod-5047 	// ********5047 
new role:  csda-ngap-inventory-to-s3-writer-cba-prod-orca 			// csda-dr-prod-1741 		// ********1741 
```

Need to attach this policy to both roles.
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "s3:PutObject",
      "Resource": "arn:aws:s3:::ss-s3-inventories-uswest2/ngap-buckets/*",
    }
  ]
}
```

**Comment by krisstanton** [comment]
UPDATE: 
BLOCKER:
I tried to create the roles on MCP and got an error.  It looks like my admin account is not authorized to create new roles.
```
Failed to create role csda-ngap-inventory-to-s3-writer-cba-prod.
User: arn:aws:sts::xxxxxxxx3323:assumed-role/mcp-tenantOperator/kstanto1 is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::xxxxxxxx3323:role/csda-ngap-inventory-to-s3-writer-cba-prod because no identity-based policy allows the iam:CreateRole action
```

**Comment by krisstanton** [comment]
WIP UPDATE:
Created Roles in CBA PROD and CBA PROD ORCA aws accounts.
These are the roles that will be handling the inventory on the NGAP side and connecting to the MCP side via assumed roles to do the writing on the MCP side.
Names:
CBA PROD: `csda-ngap-inventory-to-mcp-s3-writer-cba-prod`
CBA PROD ORCA: `csda-ngap-inventory-to-mcp-s3-writer-cba-prod-orca`
Trust Policy attached to both of the NGAP account Roles:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "sts:AssumeRole",
            "Principal": {
                "AWS": "arn:aws:iam::********3323:role/csda-ngap-inventory-to-s3-writer-cba-prod"
            }
        }
    ]
}
```

Modifying the Existing Bucket Policy on MCP Bucket `ss-s3-inventories-uswest2`
Adding this section
```
... (existing content) ...
// At the Statement Level
  {
    "Sid": "Allow NGAP (CBA PROD and ORCA) to send inventories to this bucket in sub dir ngap-buckets/",    
    "Effect": "Allow",
    "Principal": {
	"AWS": [
		"arn:aws:iam::********5047:role/csda-ngap-inventory-to-mcp-s3-writer-cba-prod",
		"arn:aws:iam::********1741:role/csda-ngap-inventory-to-mcp-s3-writer-cba-prod-orca"
	]
    },
    "Action": "s3:PutObject",
    "Resource": "arn:aws:s3:::ss-s3-inventories-uswest2/ngap-buckets/*"
  }
```

**Comment by krisstanton** [comment]
UPDATE: Inventories have been setup from all NGAP buckets, I also put the policy in place in the MCP bucket to allow the writes.  Now I need to check back after an inventory is generated and verify the destination working properly.

**Comment by krisstanton** [comment]
UPDATE: In order to make the life cycle happen for these files, it looks like there are a few options of doing this (bucket policy that targets subdirectories (if that is possible), or tracking and deleting each file after it is processed from within the DAG that reads these files)

For now, this is part of a greater discussion on how our DMS system will handle any and all lifecycle rules of the incoming inventories.  So this part of the task is shelved until that larger discussion is resolved.

**Comment by krisstanton** [comment]
The inventories are working, however this is one minor problem.
There is an extra slash being inserted into the path.
Since S3 uses key paths, this may not be a problem at all.

I made a 4th inventory to just be a test to see if some slight configuration changes will prevent this extra slash.

**Comment by hbparache** [comment]
Kris: will delete / directories
Vishal: lifecycle rules would be beneficial 
Vishal and Kris to tag up

**Comment by krisstanton** [comment]
- The sub directories with only a slash as their name have been deleted.
- The latest inventories are being sent to the correct subdirectories now (instead of a slash, the correct names are used) (Verified this)
- Met with Vishal, he is setting up the life cycle rule to point to the `ngap-buckets` subdirectory so that we just get a rolling back up of the inventory.  Older Inventory files will be deleted periodically to prevent large build up of old files.

**Comment by krisstanton** [comment]
I just set up the lifecycle rule on the ss-s3-intenvories-uswest2 bucket.
I believe I configured it correctly to only affect objects under the prefix of "ngap-buckets/"
The thing I do not know is if the targeting works correctly.
As a test, I set it to only change the storage class from "standard" to "Inteligent-Tiering" after 3 days.
If this works, and ONLY the objects under "ngap-buckets/" are affected, then I'll change this so that instead of changing the storage class, it just deletes the objects.

All this is to just to test and make sure I don't accidentally delete objects in other parts of the bucket.
